[STF-519][STF-520][STF-521] basic functionality

paramite opened the pull request:
This PR adds basic sg-agent functionality as described in spec as scheduler, executor and reactor plugins.

Depends-On: https://github.com/infrawatch/sg-core/pull/72
Depends-On: https://github.com/infrawatch/apputils/pull/23

paramite committed:
Scheduler plugin

Adds plugin for scheduling availability monitoring tasks

Depends-On: https://github.com/infrawatch/sg-core/pull/70
Implmenets: STF-519

paramite committed:
Basic CI jobs

paramite committed:
Update readme

paramite committed:
Fix copy-pasta error

vyzigold commented on README.md:
```diff
@@ -1,2 +1,20 @@
+well with other availability monitoring solutions such as sensu-core, sensu-go,
+Nagios, Icinga, Zabbix or even Datadog.
+
+TO-DO:
```
Looks to me like there is some markdown issue here. Look at the rendered readme: https://github.com/paramite/sg-agent/blob/sft-519-scheduler/README.md
It's missing newlines between the checkboxes

vyzigold commented on ci/integration/run_scheduler.sh:
```diff
@@ -0,0 +1,66 @@
+# produce some events
+/tmp/sg-core -config ./ci/integration/conf_scheduler.yaml &>/tmp/sg-core.log &
+CORE_PID=$!
+sleep 5 # should produce 5 times task1 and 2 times task2
```
This should probably work most of the time and is probably ok for CI, but a note. As far as I know sleeping usually isn't exact, so I wouldn't be surprised if sometimes the CI failed because something in the agent on here took a little longer than expected.

vyzigold replied:
I just saw you aren't checking if you got the exact amount of each task in the CI, so everything should be totally fine.

vyzigold commented on ci/integration/conf_scheduler.yaml:
```diff
@@ -0,0 +1,23 @@
+    config:
+      schedule:
+        - name: test1
+          command: "echo 'test1'"
```
I'm not sure in this exact situation, but I know sometimes it might be a security issue to execute shell commands read from a config file like this, which it looks like you are trying to do here (I think we talked about this on one of our meetings). If an attacker can somehow edit the file he could execute any script he wants and gain access to the pod, but maybe I'm just too paranoid :D. @csibbitt I think you have some background in security?

paramite replied:
That is true indeed. That's how availability monitoring frameworks work unfortunately :). Security will be handled in production by isolation of sg-agent in it's own container and restriction of execution permission via sudo config the same way as it is now handled with collectd-sensubility (executor plugin will run on overcloud nodes). In other words, by default sg-agent will be able to execute just the non-root commands + whatever you will allow it by sudo config.

pleimer replied:
Didn't we talk in the happy hour that we would rather have predefined jobs set in the executor config rather than sending arbitrary scripts over the network? I'm picturing the situation where this scheduler is running on server side sending commands to the agent on the client side. Even in a container without sudo, an attacker could execute malicious code against the OSP APIs, or post internal network info back out

pleimer replied:
In practice I do not know how often someone would be running that configuration, but we did discuss that being an option

paramite replied:
> Didn't we talk in the happy hour that we would rather have predefined jobs set in the executor config rather than sending arbitrary scripts over the network? 

We did yes, that's one of a possible use case. Right now this is not possible yet, since app plugins don't have access to transport plugins in sense that app could ask transport to send message over the bus, but it is in TODO list.

> Even in a container without sudo, an attacker could execute malicious code against the OSP APIs,

The attacker would first have to have access to the correct required API nodes. Allowing openstack command to root only and handling the command access via sudo is still possible.

> or post internal network info back out

If the attacker has access to configuration on STF server side to change check command, he probably also then have access to the whole data set being stored in there. That basically means he already have that kind of data IMO.

We can of course assume that person(s) maintaining cloud solution and it's monitoring does not have at least average IQ, but that would slip us to over-engineering simple things :) .

pleimer replied:
We don't do authentication against the AMQP bus though, so if an attacker gets access to it (say from another compromised pod), they could still spoof messages. You're right there are much bigger problems if an attacker gets root access to a container, this would just be about minimizing the damage. Chris' security updates will also solve this problem I believe. Maybe I shouldn't be the final say on security, but after further thought I think this is fine

vyzigold commented:
It's cool to see the sg-agent being born. The code looks good to me so far, there are just a few small issues I made comments about and I still need to take a look at the scheduler in apputils, which I somehow missed. Since the scheduler doesn't execute the commands and just passes them somewhere else (unless I missed something), the security "issue" isn't an issue right now and I'm OK with merging this.

paramite commented:
Yup, scheduler just requests execution. The check execution itself will be handled by execution plugin.

paramite committed:
Fix to-do list

pleimer commented on build/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+RUN echo "replace $(grep sg-core go.mod | tr '\n' ' ' | tr '\t' ' ')=> $CORE" >> go.mod
+
+# build agent plugins
+RUN ./build.sh
```
I think we want `PRODUCTION_BUILD=true ./build.sh` here to minimize the size of the sg-core image.

paramite committed:
Remove the correct invalid build script

This reverts commit f9c47d8de5d6511161e84ec926ddac665ea6f189.

pleimer commented on build/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+RUN dnf install golang git qpid-proton-c-devel -y --setopt=tsflags=nodocs
+
+# build sg-core
+RUN git clone https://github.com/infrawatch/sg-core $CORE && pushd $CORE && PRODUCTION_BUILD=true ./build.sh && popd
```
It shouldn't be much of an issue since our build process won't change often, but I was going build a makefile for sg-core and we will have to remember to update this too.

paramite replied:
Good point. Switching from bash script to makefile is a one line change, so it should not be much of an issue. And for the backward compatibility the build.sh could just call make :).

pleimer commented:
What do you think of just putting this in the sg-core repo as a core plugin rather than trying to package a separate binary that still depends on the core code base? That way there can be tighter coupling of sg-core versions and sg-agent. I understand some benefit to packaging a separate binary in that for every update in sg-core on OCP side, we don't necessarily need to release a new sg-agent. Just seems like it would be easier to keep everything in sync

paramite commented:
> What do you think of just putting this in the sg-core repo as a core plugin rather than trying to package a separate binary that still depends on the core code base? That way there can be tighter coupling of sg-core versions and sg-agent. I understand some benefit to packaging a separate binary in that for every update in sg-core on OCP side, we don't necessarily need to release a new sg-agent. Just seems like it would be easier to keep everything in sync

IMO it won't be necessary to release new sg-agent for each change in sg-core. All the protocol metadata will be in sg-agent repo, so releasing new sg-agent would be necessary only when the new sg-core feature will be needed in sg-agent data. But that is good point, that there's a value in having full featured sg-core container available on client side. The reasoning for separate repo for me was except the fact that sg-agent will be more of a client side service (with different container build process that we have on server side) was that all the sg-agent plugins will more or less share the code in lib package. This can of course be done by package imports throughout the sg-agent plugins, but that is basically creating a hidden coupling in subset of sg-core plugins.

That said, I don't mind moving the whole sg-agent bits to sg-core, it just seemed more clean to me to keep it separate.

Edit: team, please just add here thumb up if it should stay separate or thumb down if it should move to sg-core.

leifmadsen commented:
> Edit: team, please just add here thumb up if it should stay separate or thumb down if it should move to sg-core.

I don't necessarily have a strong opinion here, but packaging will be way easier if you put it in sg-core, otherwise we need to set up a whole new component for downstream release. That's not necessarily a huge deal or a blocker, and now that our release automation is much better than it used to be before, I'm definitely not nearly as scared about the amount of manual work it would have been otherwise :)

Keeping this separate might actually be better because it's going to be a component we ship with the OSP product and not with the STF product (since this is an OSP side mechanism, right?), so keeping it separate might be better?

Also, I'm not sure this is really an "STF" component per se anyways since this doesn't really rely on STF to be operational, assuming I remember and understand the architecture. The sg-agent really is a separate mechanism that can compliment STF, but does not actually require STF itself (on the OCP side) to be functional.

paramite committed:
Executor plugin

paramite committed:
Reactor logic for scheduler

paramite committed:
Reactor logic for scheduler

paramite commented:
Please ignore for now, I need to add reactor logic and end-to-end integration tests, so that you have a clear view on basic functionality. Keeping this in single PR seems more clear to me.

paramite committed:
Improve execution performace

paramite committed:
Do not execute already running task

paramite committed:
Merge branch 'refactor' into 'sft-519-scheduler'

paramite committed:
Merge typo

paramite committed:
Adapt CI

 - currently there is no integration testing, so rename accordingly
 - test also executor events

paramite committed:
Fix typo

paramite committed:
Copy pasta typo fix

paramite committed:
py2 typo fix

paramite committed:
debug

paramite committed:
Array of dicts fix

paramite committed:
Leave enough time for all tasks to execute

paramite committed:
Validation set fixes

paramite committed:
Another validation set fix

paramite committed:
Use apputils with same branch if exists

paramite committed:
Typo

paramite committed:
Use also sg-core of same branch with unit tests

paramite commented:
Sorry for the force push. Old habbits die hard.

paramite commented:
Pretty much done. Smoke tests are passing [1]. There are no unit tests yet, but I'm planning to cover the testable bits. Linting currently fails because required sg-core change is not merged yet.

[1] https://github.com/paramite/sg-agent/actions/runs/1311636282

paramite committed:
Unit tests

paramite committed:
Fix failing unit test

paramite committed:
Go1.15- compatibility fix

paramite committed:
Enable to react on metrics too

This patch adds possibility to scheduler to react on received metrics too,
so that sg-agent can sniff on external message bus commit correction steps
based on metric data.

paramite committed:
Unit test fix

paramite commented:
Let's merge this, we can change the code based on any objection later.
